examples/step-19/step-19.cc

```diff
@@ -0,0 +1,1112 @@
+ *
+ * Copyright (C) 2020 by the deal.II authors
+ *
+ * This file is part of the deal.II library.
+ *
+ * The deal.II library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE.md at
+ * the top level directory of deal.II.
+ *
+ * ---------------------------------------------------------------------
+
+ *
+ * Authors: Wolfgang Bangerth, Rene Gassmoeller, Peter Munch, 2020.
+ */
+
+
+// @sect3{Include files}
+
+// The majority of the include files used in this program are
+// well known from step-6 and similar programs:
+
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/affine_constraints.h>
+
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_refinement.h>
+
+#include <deal.II/fe/mapping_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_q.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/error_estimator.h>
+
+
+// The ones that are new are only the following three: The first declares the
+// DiscreteTime class that helps us keep track of time in a time-dependent
+// simulation. The latter two provide all of the particle functionality,
+// namely a way to keep track of particles located on a mesh (the
+// Particles::ParticleHandler class) and the ability to output these
+// particles' locations and their properties for the purposes of
+// visualization (the Particles::DataOut class).
+#include <deal.II/base/discrete_time.h>
+#include <deal.II/particles/particle_handler.h>
+#include <deal.II/particles/data_out.h>
+
+#include <fstream>
+
+using namespace dealii;
+
+
+// @sect3{Global definitions}
+
+// As is customary, we put everything that corresponds to the details of the
+// program into a namespace of its own. At the top, we define a few constants
+// for which we would rather use symbolic names than hard-coded numbers.
+//
+// Specifically, we define numbers for
+// @ref GlossBoundaryIndicator "boundary indicators"
+// for the various parts of the geometry, as well as the physical properties
+// of electrons and other specifics of the setup we use here.
+namespace Step19
+{
+  namespace BoundaryIds
+  {
+    constexpr types::boundary_id open          = 101;
+    constexpr types::boundary_id cathode       = 102;
+    constexpr types::boundary_id focus_element = 103;
+    constexpr types::boundary_id anode         = 104;
```

Comment:
Just out of curiosity, why 101, 102, 103 and 103 and not 1,2,3,4?

Reply:
I think the principle is to use numbers that are uncommon. If there are pre-defined boundary indicators previously generated by the `GridGenerator` functions, they will likely be small integers starting from zero, but not in this rather randomly chosen range. It just avoids the possibility for conflicts, and also reduces the temptation to just spell these numbers out (because you will probably never remember which is which, whereas you might have been tempted if they had started at 0).

It's a good enough question that I copied something like the text above into the program itself.